### PR TITLE
Allow binary operators to be used as lambdas

### PR DESCRIPTION
**Pull request by ghallak**
Fixes: #228

### CODE REVIEW

File: test/contracts/operator_lambdas.aes [hunk 1]
```diff
@@ -0,0 +1,14 @@
+include "List.aes"
+
+contract C =
+    type state = int
+
+    function sum(l : list(bool)) : bool = List.foldl((||), false, l)
```
Comment (radrow) [inline test/contracts/operator_lambdas.aes:6]: sum? :p

File: src/aeso_parser.erl [hunk 2]
```diff
@@ -472,6 +472,19 @@ id()       -> token(id).
 tvar()     -> token(tvar).
 str()      -> token(string).
 
+binop_as_lam() ->
+    BinOps = ['&&', '||',
```
Comment (radrow) [inline src/aeso_parser.erl:476]: It's sorta awkward that we maintain yet another lists of operators (one in lexer, one in expression parser and this one here)... Maybe there is a nicer way of doing so, so they won't desync?

Also, you don't want to allow `..` operator here, it doesn't make any sense to make it a lambda

**Review by radrow — APPROVED** [review]

File: test/contracts/operator_lambdas.aes (hunk 1 above)
Comment (ghallak) [reply test/contracts/operator_lambdas.aes:6]: Will change this to `any`.

File: src/aeso_ast_infer_types.erl [hunk 3]
```diff
@@ -1641,7 +1641,10 @@ infer_expr(Env, {app, Ann, Fun, Args0} = App) ->
             GeneralResultType = fresh_uvar(Ann),
             ResultType = fresh_uvar(Ann),
             when_warning(warn_unused_functions,
-                         fun() -> register_function_call(Namespace ++ qname(CurrentFun), Name) end),
+                         fun() -> if element(1, Name) == lam -> ok;
+                                     true -> register_function_call(Namespace ++ qname(CurrentFun), Name)
```
[the remark this answers is not in the document]
Comment (ghallak) [reply src/aeso_ast_infer_types.erl:1645]: This was an existing bug that nobody has noticed. It didn't allow lambdas to be called directly on arguments (e.g. you can't do `((x) => x))(1)`).

I've noticed this when I was trying to do `(+)(1, 2)` and fixed it.

File: test/contracts/operator_lambdas.aes (hunk 1 above)
Comment (ghallak) [reply test/contracts/operator_lambdas.aes]: Changed it to `any` here 642f5a4da32090f2e1dd277b9d9b84482b3b8942

File: src/aeso_parser.erl (hunk 2 above)
Comment (ghallak) [reply src/aeso_parser.erl:476]: Disabled `..` here 8851629fc7f0b1d8599c6f30166b0437857d570b

**Review by hanssv — APPROVED** [review]

**Pull request merged**